### PR TITLE
Move flattening/unflattening logic of JIT to C

### PR DESCRIPTION
**Pull request by apaszke**
BatchNorm tests are broken, because they're passing tensors into closures. We should figure out a better story for batch norm anyway, so I just disabled the tests for now.

I also took the chance and simplified parts of our JIT code

### CODE REVIEW

File: torch/csrc/jit/python_arg_parser.cpp [hunk 1]
```diff
@@ -0,0 +1,119 @@
+#include "python_arg_parser.h"
+
+namespace torch { namespace jit { namespace python {
+
+struct ParsedArgs {
+  std::vector<py::handle> vars;
+  std::string desc;
```
Comment (ezyang) [inline torch/csrc/jit/python_arg_parser.cpp]: It would be nice if you could add a comment saying what the format of this string is.

**Comment by ezyang** [comment]
Nice!! Do you know what the timings are?

**Comment by apaszke** [comment]
Yes, in `benchmark/scripts/lstm.py` with flag `--jit` the time drops from around 204us to 164us. Time spent on flattening/building up in_key drops from >40us to 6us. Unflattening is slightly faster, but it didn't seem to be such a significant bottleneck before.

I'll add the comments later

File: torch/csrc/jit/python_arg_parser.cpp [hunk 2]
```diff
@@ -0,0 +1,145 @@
+#include "python_arg_parser.h"
```
Comment (colesbury) [inline torch/csrc/jit/python_arg_parser.cpp]: Can you name this something different? Files with the same name in different directories get confusing.

Comment (apaszke) [reply torch/csrc/jit/python_arg_parser.cpp]: Sure. Sorry, I forgot we already have one like this

Comment (apaszke) [reply torch/csrc/jit/python_arg_parser.cpp]: I renamed them to `python_arg_flatten.*`

**Comment by zdevito** [comment]
Looks generally good, with a great improvement to building in_key. I couldn't find where the tensor sizes are encoded in the key in the new version. Are we no longer using that info or am I just missing the line that does it?

**Comment by apaszke** [comment]
Yeah I just realized that we no longer check the sizes and strides. I'll fix that in a PR that moves more of tracing logic to C++

**Comment by dzhulgakov** [comment]
Hey folks, just breaking batch_norm export (and disabling tests) is not cool as it completely breaks ONNX export for anything having batchnorm!

In fact, since we're doing match by the number of inputs when we're exporting the initializers in the graph (https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/export.cpp#L221) I suspect it also breaks any export if some of the modules has any buffers (since we compare the order of state_dict and parameters())

Is there a more plausible fix for buffers handling? One option might be to just override the symbolic export for the entire module (batchnorm) that has buffers but that would require hack similar to https://github.com/pytorch/pytorch/pull/3044/

@apaszke @ezyang @houseroad - any suggestions?

**Comment by apaszke** [comment]
Oops, my bad. I didn't break the onnx pass on purpose, the test I disabled was meant to be a JIT check, and we don't intend to be able to re-execute batch norm ops yet. I'll add back support for tensors today in the evening.

**Comment by apaszke** [comment]
(This will only be a hotfix. We should figure out a more robust solution.)

**Comment by apaszke** [comment]
Actually, once Variables and Tensors will be fused, we won't have a clear distinction between parameters and buffers and this will simplify the code a lot (e.g. BatchNorm will be able to take buffers as inputs)

**Comment by apaszke** [comment]
Fixed in #3691
